typed-protocols/src/Network/TypedProtocol/Core.hs

```diff
@@ -0,0 +1,355 @@
+-- outgoing messages.
+--
+type WeHaveAgency   (pk :: PeerKind) st = PeerHasAgency             pk  st
+
+-- | A synonym for an state token in which the other peer has agency. This is
+-- parametrised over the client or server roles. In either case the other peer
+-- has agency.
+--
+-- This shows up when we are receiving messages, or dealing with decoding
+-- incoming messages.
+--
+type TheyHaveAgency (pk :: PeerKind) st = PeerHasAgency (FlipAgency pk) st
+
+-- | A type function to flip the client and server roles.
+--
+type family FlipAgency (pk :: PeerKind) where
+  FlipAgency AsClient = AsServer
+  FlipAgency AsServer = AsClient
+
+
+-- | A description of a peer that engages in a protocol.
+--
+-- The protocol describes what messages peers /may/ send or /must/ accept.
+-- A particular peer implementation decides what to actually do within the
+-- constraints of the protocol.
+--
+-- Peers engage in a protocol in either the client or server role. Of course
+-- the client role can only interact with the serve role for the same protocol
+-- and vice versa.
+--
+-- 'Peer' has several type arguments:
+--
+-- * the protocol itself;
+-- * the client\/server role;
+-- *.the current protocol state;
+-- * the monad in which the peer operates; and
+-- * the type of any final result once the peer terminates.
+--
+-- For example:
+--
+-- > pingPongClientExample :: Int -> Peer PingPong AsClient StIdle m ()
+-- > pingPongServerExample ::        Peer PingPong AsServer StIdle m Int
+--
+-- The actions that a peer can take are:
+--
+-- * to perform local monadic effects
+-- * to terminate with a result (but only in a terminal protocol state)
+-- * to send a message (but only in a protocol state in which we have agency)
+-- * to wait to receive a message (but only in a protocol state in which the
+--   other peer has agency)
+--
+-- In the 'Done', 'Yield' and 'Await' cases we must provide evidence of both
+-- the protocol state we are in and that the appropriate peer has agency.
+-- This takes the form of 'ClientAgency' or 'ServerAgency' applied to a
+-- protocol-specific state token: either a 'ClientHasAgency' or a
+-- 'ServerHasAgency' token for the protocol. The 'Done' state does not need
+-- the extra agency information.
+--
+-- While this evidence must be provided, the types guarantee that it is not
+-- possible to supply incorrect evidence.
+--
+data Peer ps (pk :: PeerKind) (st :: ps) m a where
+
+  -- | Perform a local monadic effect and then continue.
+  --
+  -- Example:
+  --
+  -- > Effect $ do
+  -- >   ...          -- actions in the monad
+  -- >   return $ ... -- another Peer value
+  --
+  Effect :: m (Peer ps pk st m a)
+         ->    Peer ps pk st m a
+
+  -- | Terminate with a result. A state token must be provided from the
+  -- 'NobodyHasAgency' states, so show that this is a state in which we can
+  -- terminate.
+  --
+  -- Example:
+  --
+  -- > Done TokDone result
```

Comment:
Maybe add another example from ping pong:
```
Yield (ClientAgency TokIdle) MsgTerminate (Done TokDOne result)
```
since usually this will be used in this way.